File: fpga-xrt/src/ffi.cc
#include "src/ffi.h"
// #include "fpga-xrt/src/ffi.rs.h"

#include <iostream>

// Xclbin
namespace xrt
{
  rust::Str XclbinArg::name() const
  {
    return get_name();
  }

  std::unique_ptr<std::vector<XclbinMem>> XclbinArg::mems() const
  {
    std::vector<xclbin::mem> input = get_mems();
    std::vector<XclbinMem> output(input.begin(), input.end());
    return std::make_unique<std::vector<XclbinMem>>(output);
  }

  rust::Str XclbinKernel::name() const
  {
    return get_name();
  }

  rust::Str XclbinIp::name() const
  {
    return get_name();
  }

  std::unique_ptr<std::vector<XclbinArg>> XclbinIp::args() const
  {
    std::vector<xclbin::arg> input = get_args();
    std::vector<XclbinArg> output(input.begin(), input.end());
    return std::make_unique<std::vector<XclbinArg>>(output);
  }

  std::unique_ptr<XclbinArg> XclbinIp::arg(int32_t index) const
  {
    xclbin::arg arg = get_arg(index);
    return std::make_unique<XclbinArg>(arg);
  }

  std::unique_ptr<Xclbin> new_xclbin(const rust::Slice<const int8_t> data)
  {
    std::vector<char> input(data.begin(), data.end());
    return std::make_unique<Xclbin>(input);
  }

  std::unique_ptr<std::vector<XclbinKernel>> Xclbin::kernels() const
  {
    std::vector<xclbin::kernel> input = get_kernels();
    std::vector<XclbinKernel> output(input.begin(), input.end());
    return std::make_unique<std::vector<XclbinKernel>>(output);
  }

  std::unique_ptr<XclbinKernel> Xclbin::kernel(const rust::Str name) const
  {
    return std::make_unique<XclbinKernel>(get_kernel(std::string(name)));
  }

  std::unique_ptr<std::vector<XclbinIp>> Xclbin::ips() const
  {
    std::vector<xclbin::ip> input = get_ips();
    std::vector<XclbinIp> output(input.begin(), input.end());
    return std::make_unique<std::vector<XclbinIp>>(output);
  }

  std::unique_ptr<XclbinIp> Xclbin::ip(const rust::Str name) const
  {
    return std::make_unique<XclbinIp>(get_ip(std::string(name)));
  }

  rust::Str Xclbin::xsa_name() const
  {
    return get_xsa_name();
  }

  std::array<unsigned char, 16> Xclbin::uuid() const
  {
    std::array<unsigned char, 16> array;
    memcpy(array.data(), xclbin::get_uuid().get(), sizeof(unsigned char) * 16);
    return array;
  }

}

// Device
namespace xrt
{
  std::unique_ptr<Device> new_device(unsigned int index)
  {
    return std::make_unique<Device>(index);
  }

  std::unique_ptr<Device> new_device_bdf(const rust::Str bdf)
  {
    return std::make_unique<Device>(std::string(bdf));
  }

  rust::Str Device::bdf() const
  {
    return get_info<xrt::info::device::bdf>();
  }

  std::array<unsigned char, 16> Device::interface_uuid() const
  {
    std::array<unsigned char, 16> array;
    memcpy(array.data(), get_info<xrt::info::device::interface_uuid>().get(), sizeof(unsigned char) * 16);
    return array;
  }

  uint32_t Device::kdma() const
  {
    return get_info<xrt::info::device::kdma>();
  }

  unsigned long Device::max_clock_frequency_mhz() const
  {
    return get_info<xrt::info::device::max_clock_frequency_mhz>();
  }

  bool Device::m2m() const
  {
    return get_info<xrt::info::device::m2m>();
  }

  rust::Str Device::name() const
  {
    return get_info<xrt::info::device::name>();
  }

  bool Device::nodma() const
  {
    return get_info<xrt::info::device::nodma>();
  }

  bool Device::offline() const
  {
    return get_info<xrt::info::device::offline>();
  }

  // rust::String Device::electrical() const
  // {
  //   return get_info<xrt::info::device::electrical>();
  // }

  // rust::String Device::thermal() const
  // {
  //   return get_info<xrt::info::device::thermal>();
  // }

  // rust::String Device::mechanical() const
  // {
  //   return get_info<xrt::info::device::mechanical>();
  // }

  // rust::String Device::memory() const
  // {
  //   return get_info<xrt::info::device::memory>();
  // }

  // rust::String Device::platform() const
  // {
  //   return get_info<xrt::info::device::platform>();
  // }

  // rust::String Device::pcie_info() const
  // {
  //   return get_info<xrt::info::device::pcie_info>();
  // }

  // rust::String Device::host() const
  // {
  //   return get_info<xrt::info::device::host>();
  // }

  // rust::String Device::dynamic_regions() const
  // {
  //   return get_info<xrt::info::device::dynamic_regions>();
  // }

  std::array<unsigned char, 16> Device::xclbin_uuid() const
  {
    std::array<unsigned char, 16> array;
    memcpy(array.data(), get_xclbin_uuid().get(), sizeof(unsigned char) * 16);
    return array;
  }

  std::array<unsigned char, 16> Device::load(const Xclbin &xclbin)
  {
    auto uuid = load_xclbin(xclbin).get();
    std::array<unsigned char, 16> array;
    memcpy(array.data(), uuid, sizeof(unsigned char) * 16);
    return array;
  }
}

// Kernel
namespace xrt
{
  std::unique_ptr<kernel> new_kernel(const Device &device,
                                     std::array<unsigned char, 16> xclbin_id,
                                     const rust::Str name,
                                     kernel_cu_access_mode mode)
  {
    xrt::uuid uuid(xclbin_id.data());
    // // workaround that registers the alxf
    // std::shared_ptr<xrt_core::device> handle = device.get_handle();
    // handle->register_axlf(xclbin.get_axlf());
    // xrt::xclbin bin(xclbin);
    return std::make_unique<kernel>(device, uuid, std::string(name), mode);
  }
}

// IP
namespace xrt
{
  std::unique_ptr<ip> new_ip(const Device &device,
                             std::array<unsigned char, 16> xclbin_id,
                             const rust::Str name)
  {
    xrt::uuid uuid(xclbin_id.data());
    return std::make_unique<ip>(device, uuid, std::string(name));
  }
}

// Ini
namespace xrt
{
  void set_ini(rust::Str key, rust::Str value)
  {
    xrt::ini::set(std::string(key), std::string(value));
  }
}

File: fpga-xrt/src/ffi.rs
#[cxx::bridge]
mod ffi {

    // xrt_device.h
    #[namespace = "xrt"]
    unsafe extern "C++" {
        include!("src/ffi.h");

        type Device;
        fn new_device(index: u32) -> Result<UniquePtr<Device>>;
        fn new_device_bdf(bdf: &str) -> Result<UniquePtr<Device>>;

        // Modified members
        fn xclbin_uuid(self: &Device) -> [u8; 16];
        fn load(self: Pin<&mut Device>, xclbin: &Xclbin) -> [u8; 16];

        // Custom members
        fn bdf(self: &Device) -> &str;
        fn interface_uuid(self: &Device) -> [u8; 16];
        fn kdma(self: &Device) -> u32;
        fn max_clock_frequency_mhz(self: &Device) -> u64;
        fn m2m(self: &Device) -> bool;
        fn name(self: &Device) -> &str;
        fn nodma(self: &Device) -> bool;
        fn offline(self: &Device) -> bool;
        // fn electrical(self: &Device) -> String;
        // fn thermal(self: &Device) -> String;
        // fn mechanical(self: &Device) -> String;
        // fn memory(self: &Device) -> String;
        // fn platform(self: &Device) -> String;
        // fn pcie_info(self: &Device) -> String;
        // fn host(self: &Device) -> String;
        // fn dynamic_regions(self: &Device) -> String;
    }

    // xrt_xclbin.h
    #[namespace = "xrt"]
    unsafe extern "C++" {
        include!("src/ffi.h");

        type XclbinArg;
        fn name(self: &XclbinArg) -> &str;
        fn mems(self: &XclbinArg) -> UniquePtr<CxxVector<XclbinMem>>;
        
        type XclbinIp;
        fn name(self: &XclbinIp) -> &str;
        #[rust_name = "num_args"]
        fn get_num_args(self: &XclbinIp) -> usize;
        fn args(self: &XclbinIp) -> UniquePtr<CxxVector<XclbinArg>>;
        fn arg(self: &XclbinIp, index: i32) -> UniquePtr<XclbinArg>;
        #[rust_name = "base_address"]
        fn get_base_address(self: &XclbinIp) -> u64;

        type XclbinKernel;
        fn name(self: &XclbinKernel) -> &str;
        
        type XclbinMem;

        type Xclbin;
        fn new_xclbin(bytes: &[i8]) -> Result<UniquePtr<Xclbin>>;
        fn kernels(self: &Xclbin) -> UniquePtr<CxxVector<XclbinKernel>>;
        fn kernel(self: &Xclbin, name: &str) -> UniquePtr<XclbinKernel>;
        fn ips(self: &Xclbin) -> UniquePtr<CxxVector<XclbinIp>>;
        fn ip(self: &Xclbin, name: &str) -> UniquePtr<XclbinIp>;
        fn xsa_name(self: &Xclbin) -> &str;
        fn uuid(self: &Xclbin) -> [u8; 16];
    }

    #[namespace = "xrt"]
    enum kernel_cu_access_mode {
        exclusive,
        shared,
        none,
    }

    // xrt_kernel.h
    #[namespace = "xrt"]
    unsafe extern "C++" {
        include!("src/ffi.h");
        
        type kernel_cu_access_mode;

        type kernel;
        fn new_kernel(
            device: &Device,
            xclbin_id: [u8; 16],
            name: &str,
            cu_access_mode: kernel_cu_access_mode,
        ) -> Result<UniquePtr<kernel>>;
        fn group_id(self: &kernel, argno: i32) -> i32;
        fn offset(self: &kernel, argno: i32) -> u32;
        fn read_register(self: &kernel, offset: u32) -> u32;
        fn write_register(self: Pin<&mut kernel>, offset: u32, data: u32) -> Result<()>;
    }

    // xrt_ip.h
    #[namespace = "xrt"]
    unsafe extern "C++" {
        include!("src/ffi.h");
        
        type ip;
        fn new_ip(
            device: &Device,
            xclbin_id: [u8; 16],
            name: &str
        ) -> Result<UniquePtr<ip>>;
        fn read_register(self: &ip, offset: u32) -> u32;
        fn write_register(self: Pin<&mut ip>, offset: u32, data: u32) -> Result<()>;
    }

    // xrt_ini.h
    #[namespace = "xrt"]
    unsafe extern "C++" {
        include!("src/ffi.h");

        fn set_ini(key: &str, value: &str) -> Result<()>;
    }
}

pub use ffi::*;
